Microsoft.Toolkit.Uwp.UI.Controls.DataGrid/DataGrid/DataGridRows.cs

```diff
@@ -2784,6 +2784,9 @@ private void ScrollSlotsByHeight(double height)
                 double newVerticalOffset = _verticalOffset + height;
                 if (height > 0)
                 {
+                    // Fire operation for loading more data from incremental data source
+                    LoadMoreDataFromIncrementalItemsSourceAsync();
```

Comment:
Filip - I have been doing some testing and came to the conclusion that this is the wrong thing to do.

LoadMoreDataFromIncrementalItemsSource() should be called when either VerticalOffset increased, or CellsHeight increased, or EdgedRowsHeightCalculated decreased.

The problems with the code above is that it's only dealing with the situation where VerticalOffset increases, and in fact it's using the old value _verticalOffset instead of the new one.

So I came up with this approach instead:

1. Slip up LoadMoreDataFromIncrementalItemsSource() into two:

```
    private void LoadMoreDataFromIncrementalItemsSource()
    {
        LoadMoreDataFromIncrementalItemsSource(totalVisibleHeight: EdgedRowsHeightCalculated);
    }

    private void LoadMoreDataFromIncrementalItemsSource(double totalVisibleHeight)
    {
        if (IncrementalLoadingTrigger == IncrementalLoadingTrigger.Edge &&
            DataConnection.IsDataSourceIncremental &&
            DataConnection.HasMoreItems &&
            !DataConnection.IsLoadingMoreItems)
        {
            var bottomScrolledOffHeight = Math.Max(0, totalVisibleHeight - CellsHeight - VerticalOffset);

            if ((IncrementalLoadingThreshold * CellsHeight) >= bottomScrolledOffHeight)
            {
                var numberOfRowsToLoad = Math.Max(1, (int)(DataFetchSize * CellsHeight / RowHeightEstimate));

                DataConnection.LoadMoreItems((uint)numberOfRowsToLoad);
            }
        }
    }

```

2. Undo the changes to ScrollSlotsByHeight(double height).

3. Add a setter to VerticalOffset:

```
    set
    {
        bool loadMoreDataFromIncrementalItemsSource = _verticalOffset < value;

        _verticalOffset = value;

        if (loadMoreDataFromIncrementalItemsSource)
        {
            LoadMoreDataFromIncrementalItemsSource();
        }
    }

``` 

4. Wherever there is a _verticalOffset = someValue, replace it with VerticalOffset = someValue.
This takes care of the case where VerticalOffset increases.

5. Replace the existing line 
`        _rowsPresenterAvailableSize = value;
`
with
```
            bool loadMoreDataFromIncrementalItemsSource =
                _rowsPresenterAvailableSize.HasValue &&
                value.HasValue &&
                value.Value.Height > _rowsPresenterAvailableSize.Value.Height;

             _rowsPresenterAvailableSize = value;

            if (loadMoreDataFromIncrementalItemsSource)
            {
                LoadMoreDataFromIncrementalItemsSource();
            }

```
This takes care of the case where CellsHeight increases.

6. The EdgedRowsHeightCalculated getter needs some changes too:

```
  private double EdgedRowsHeightCalculated
  {
      get
      {
          // If we're not displaying any rows or if we have infinite space the, relative height of our rows is 0
          if (this.DisplayData.LastScrollingSlot == -1 || double.IsPositiveInfinity(this.AvailableSlotElementRoom))
          {
              if (_oldEdgedRowsHeightCalculated > 0)
              {
                  _oldEdgedRowsHeightCalculated = 0;
                  LoadMoreDataFromIncrementalItemsSource(0);
              }

              return 0;
          }

          ...
          double totalDetailsHeight = detailsCount * this.RowDetailsHeightEstimate;
          double newEdgedRowsHeightCalculated = totalRowsHeight + totalDetailsHeight;
          bool loadMoreDataFromIncrementalItemsSource = newEdgedRowsHeightCalculated < _oldEdgedRowsHeightCalculated;

          _oldEdgedRowsHeightCalculated = newEdgedRowsHeightCalculated;

          if (loadMoreDataFromIncrementalItemsSource)
          {
              LoadMoreDataFromIncrementalItemsSource(newEdgedRowsHeightCalculated);
          }

          return newEdgedRowsHeightCalculated;
      }
  }

```
_oldEdgedRowsHeightCalculated is a new 'double' field for DataGrid that is initialized to 0.
This takes care of the case where EdgedRowsHeightCalculated decreases.

Hope this makes all sense. Thanks.

Reply:
@RBrid Just for my understanding regarding point 6 above, what is the expected behavior when the EdgeRowsHeightCalculated decreases?  How is it decreased (what kind of scenario), and why do we want to load more data when it decreases?

Reply:
@RBrid When testing I ran in to some issues with step 6 above. When clearing a collection bound to the ItemsSource, it jumps directly into
```
if (_oldEdgedRowsHeightCalculated > 0)
{
   _oldEdgedRowsHeightCalculated = 0;
   LoadMoreDataFromIncrementalItemsSource(0);
}
```
and starts loading data from the last known position which then is displayed at the top of the grid. I don't believe it is a desired behavior load data automatically clearing a collection. In my opinion it should be told to load more data explicitly. Also, loading data from the last known position where data was retrieved from cannot be desirable. If anything, it should start loading from the beginning.

When commenting out the line ` LoadMoreDataFromIncrementalItemsSource(0);`, I can clear the collection and it works as expected.

Making sure that the ItemsSource actually has data before starting to load also worked as below:
```
if (DataConnection.DataSource != null && DataConnection.DataSource.GetEnumerator().MoveNext())
{
   LoadMoreDataFromIncrementalItemsSource(0);
}
```

Is the solution above acceptable? If not, how would you prefer to solve this issue?

Reply:
@filipwa84, I looked into this and believe that the condition "if (DataConnection.DataSource != null && DataConnection.DataSource.GetEnumerator().MoveNext())" needs to be removed.
Clearing the collection is similar to removing each item one by one. Whether you remove them one by one or all at ones, this causes the EdgedRowsHeightCalculated value to decrease and should cause a call to LoadMoreDataFromIncrementalItemsSource(0), because the trigger criteria may now be met. 

LoadMoreDataFromIncrementalItemsSource(double) will check if HasMoreItems is True or False.
If it's True, a new load may be requested. Otherwise none is requested.

I suspect your implementation of ISupportIncrementalLoading.HasMoreItems returns True when you expected it to return False.

Also regarding the 'load starting index', ISupportIncrementalLoading has no notion of that. It's its implementation that may have one.

Hope this makes sense.  Let me know. Thank you.

Reply:
@RBrid You were right, changing the ISupportIncrementalLoading implementation to explicitly set `HasMoreItems = false` when the collection is cleared solved the issue.

Will change it back.